src/celeritas/track/detail/TrackSortUtils.hh

```diff
@@ -54,10 +54,31 @@ void shuffle_track_slots<MemSpace::device>(
 
 //---------------------------------------------------------------------------//
 // Sort or partition tracks
-
 void sort_tracks(HostRef<CoreStateData> const&, TrackOrder);
 void sort_tracks(DeviceRef<CoreStateData> const&, TrackOrder);
 
+//---------------------------------------------------------------------------//
+// Count tracks associated to each action
+template<MemSpace M>
+void count_tracks_per_action(
+    CoreStateData<Ownership::reference, M> const&,
+    Span<ThreadId>,
+    Collection<ThreadId, Ownership::value, MemSpace::host, ActionId>&,
+    TrackOrder);
+
+template<>
+void count_tracks_per_action<MemSpace::host>(
+    HostRef<CoreStateData> const&,
+    Span<ThreadId>,
+    Collection<ThreadId, Ownership::value, MemSpace::host, ActionId>&,
+    TrackOrder);
+template<>
+void count_tracks_per_action<MemSpace::device>(
+    DeviceRef<CoreStateData> const&,
+    Span<ThreadId>,
+    Collection<ThreadId, Ownership::value, MemSpace::host, ActionId>&,
+    TrackOrder);
+
```

Comment:
Not sure what the best API would be here between 1) collection as output parameter 2) extract the back-filling part to a separate function 3) return a collection.